### PR TITLE
[improvement](index) support CANCEL BUILD INDEX

### PR DESCRIPTION
**Pull request by xiaokang**
## Proposed changes

Issue Number: close #xxx

Currently, BUILD INDEX can not be canceled. This pr add CANCEL BUILD INDEX as follows:
`CANCEL BUILD INDEX index_name ON table_name [(jobid1,jobid2)]`

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xiaokang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.27 seconds
 stream load tsv:          534 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.8 seconds inserted 10000000 Rows, about 347K ops/s
 storage size: 17162167032 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.15 seconds
 stream load tsv:          529 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.7 seconds inserted 10000000 Rows, about 348K ops/s
 storage size: 17161978221 Bytes

**Comment by xiaokang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.59 seconds
 stream load tsv:          531 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17162210008 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/alter/IndexChangeJob.java [hunk 1]
```diff
@@ -321,10 +322,31 @@ protected void runRunningJob() throws AlterCancelException {
...
+        if (jobState.isFinalState()) {
+            return false;
+        }
+
+        cancelInternal();
+
+        jobState = JobState.CANCELLED;
+        this.errMsg = errMsg;
+        this.finishedTimeMs = System.currentTimeMillis();
+        LOG.info("cancel index job {}, err: {}", jobId, errMsg);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/alter/IndexChangeJob.java:339]: print log after `Env.getCurrentEnv().getEditLog().logIndexChangeJob(this);`

File: fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java [hunk 2]
```diff
@@ -2370,6 +2378,69 @@ public void cancel(CancelStmt stmt) throws DdlException {
...
+        Preconditions.checkState(!Strings.isNullOrEmpty(dbName));
+        Preconditions.checkState(!Strings.isNullOrEmpty(tableName));
+
+        Database db = Env.getCurrentInternalCatalog().getDbOrDdlException(dbName);
+
+        List<IndexChangeJob> jobList = new ArrayList<>();
+
+        OlapTable olapTable;
+        try {
+            olapTable = (OlapTable) db.getTableOrMetaException(tableName, Table.TableType.OLAP);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:2393]: You can use `getTableOrException`

File: fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java [hunk 3]
```diff
@@ -2370,6 +2378,69 @@ public void cancel(CancelStmt stmt) throws DdlException {
...
+            olapTable.writeUnlock();
+        }
+
+        // alter job v2's cancel must be called outside the table lock
+        if (jobList.size() > 0) {
+            for (IndexChangeJob job : jobList) {
+                long jobId = job.getJobId();
+                LOG.debug("cancel build index job {} on table {}", jobId, tableName);
+                if (!job.cancel("user cancelled")) {
+                    LOG.info("cancel build index job {} on table {} failed", jobId, tableName);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:2433]:
```suggestion
                    LOG.warn("cancel build index job {} on table {} failed", jobId, tableName);
```

File: fe/fe-core/src/main/java/org/apache/doris/alter/IndexChangeJob.java (hunk 1 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/alter/IndexChangeJob.java:339]: Missing `replayCancelled()`

**Comment by xiaokang** [comment]
run buildall

**Comment by xiaokang** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.52 seconds
 stream load tsv:          529 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162143250 Bytes

**Review by qidaye — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Review by wm1581066 — APPROVED** [review]
LGTM

**Pull request merged**